docs/proposals/20201125-jaeger-traces-store.md

```diff
@@ -0,0 +1,124 @@
+
+#### Single Jaeger instance
+
+This deployment topology deploys a single Jaeger instance for all tenants with a single Elasticsearch instance. The data for each tenant would contain a unique label identifying a tenant. The label would be dynamically injected in the Observatorium API service or in the OpenTelemetry collector.
+
+This deployment strategy will not work with the current Jaeger components. Here are few issues. The list might not be complete:
+* Service and operation name API. The service query API nor storage does not expose/store labels.
+* Dependency/service architecture diagram. To support this feature dependency schema, query, collector and Spark aggregation would have to change.
+* Data retention/TTL configurable pre tenant. The data retention in Elasticsearch is configurable per index.
+
+##### OpenTelemetry collector configuration
+
+The OpenTelemetry collector is configured with a single exporter to export data to a single Jaeger collector.
+
+#### Jaeger instance per tenant
+
+This deployment topology deploys a Jaeger instance per tenant and all Jaeger instances use a single Elasticsearch instance. The data for each tenant is stored in a separate Elasticsearch index. Jaeger was designed to separate tenant data per index. Therefore, all Jaeger functionality works well with this deployment strategy.
+
+##### OpenTelemetry collector configuration
+
+The OpenTelemetry collector is configured with an exporter per tenant that exports data to Jaeger collector allocated per a single tenant.
+
+### Jaeger Query
+
+There are two ways users could access data stored in Jaeger:
+1. Expose Jaeger Query API in the Observatorium API.
+2. Expose Jaeger UI outside the cluster.
+
+#### Expose Jaeger Query API
+
+Jaeger query API would be exposed in the Observatorium API. The users would have to deploy Jaeger UI externally and use Observatorium API as a data source.
+
+There are multiple Jaeger query APIs, in this case we propose to expose [api_v3](https://github.com/jaegertracing/jaeger-idl/tree/master/proto/api_v3). This API returns OpenTelemetry compatible payload.
+
+#### Expose Jaeger UI
+
+Jaeger Query with UI would be deployed in the Observatorium cluster and exposed outside the cluster. Each tenant would get a unique URL to access Jaeger Query UI.
+
+## Action Plan
+
+Given the issues with the single Jaeger instance topology, the best approach is to deploy a Jaeger instance per tenant that separates tenant's data by using different indices. This approach will require adding a configuration if a new tenant is added to the system.
```

Comment:
Could we in the section above make it a bit more clear which solution is the chosen one and which one is the alternative that will not be implemented? Right now it seems like both options will be considered.

Reply:
I understand the issues with single Jaeger instance variation, but I'm wondering:
1) How will we be able to ensure seamless addition of new tenants? It looks like this would require a concerted effort of first adding a new Jaeger instance and only then adding tenant to the API. I'm wondering if we can improve on this in any way to make the experience more coherent.
2) Will this solution scale without issues for e.g. hundreds of tenants, meaning hundreds of Jaeger instances? Are there any performance or operation overhead concerns?

Reply:
>  It looks like this would require a concerted effort of first adding a new Jaeger instance and only then adding tenant to the API.

As far as I know the API does not need any changes if a new tenant is added and this should be case with Jaeger as well. However, to add a new tenant Jaeger deployment will have to change (e.g. add a new collector).

Reply:
At minimum, adding tenant requires to update the tenant configuration (see [example](https://github.com/observatorium/api/blob/main/examples/tenants/two.yaml)) and this currently necessitates to bump any running instances of API to make the new config take effect. My point was regarding that fact exactly - the user will need to first bump API and then introduce the deployment change on the tracing backend (or rather vice versa). But I'm not sure if with the 'one tenant per instance' model we can improve on this.